### PR TITLE
Verilator

### PR DESCRIPTION
**Pull request by powerjg**

### CODE REVIEW

File: configs/example/dinocpu.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+# Set up the binary to load
+system.kernel = binary
+system.system_port = system.mem_ctrl.port
+
+# Create the dinocpu verilator wrapper
+system.dinocpu = VerilatorDinoCPU()
+
+# Create the black box memory wrapper
+system.memblackbox = VerilatorMemBlackBox()
```
Comment (powerjg) [inline configs/example/dinocpu.py]: The DINO CPU probably needs a pointer to this.

File: configs/example/dinocpu.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+system.memblackbox.instPort = system.mem_ctrl.port
+system.memblackbox.dataPort = system.mem_ctrl.port
+
+system.dinocpu.latency = '1ns'
```
Comment (powerjg) [inline configs/example/dinocpu.py]: Should this be a clock period instead of latency?

File: src/verilator/SConscript [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Import('*')
```
Comment (powerjg) [inline src/verilator/SConscript]: FYI, we need licenses on all of these files. See other files for examples. The copyright is to University Regents of California or something like that. The DArchR/docs repo has the exact wording.

File: src/verilator/VerilatorDinoCPU.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#include "VerilatorDinoCPU.hh"
+#include "base/logging.hh"
+#include "sim/sim_exit.hh"
+
+VerilatorDinoCPU::VerilatorDinoCPU(VerilatorDinoCPUParams *params) :
+    SimObject(params),
+    event([this]{updateCycle();}, params -> name),
+    latency(params -> latency),
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.cc]: I've never seen spaces around the `->` operator before. I don't think you should do that. It would be like spaces around the `.` operator. `cpu . memory` doesn't look right :)

File: src/verilator/VerilatorDinoCPU.hh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+#include "params/VerilatorDinoCPU.hh"
+#include "sim/sim_object.hh"
+
+class VerilatorDinoCPU
+{
+    private:
+        void updateCycle();
+
+        VTop top;
+        VerilatorMemBlackBox verilatorMem;
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.hh]: Should probably be a pointer.

Also, to get this pointer into this object, you can add a parameter to the SimObject description file that requires VerilatorMemBlackBox object.

File: src/verilator/VerilatorDinoCPU.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from m5.params import *
+from m5.SimObject import SimObject
+
+class VerilatorDinoCPU(SimObject):
+    type = 'VerilatorDinoCPU'
+    cxx_header = "verilator/VerilatorDinoCPU.hh"
+
+    latency = Param.Latency("Clock period of device under test")
+    stages = Param.Int(1, "Number of stages in the device under test")
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: Here, you probably also want

```
memory_black_box = Param.VerilatorMemBlackBox("The memory to connect between verilator and gem5")
```

File: src/verilator/verilator.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import m5
```
Comment (powerjg) [inline src/verilator/verilator.py]: Should this be in the configs directory?? I think so!

**Review by powerjg — COMMENTED** [review]
Looks like we're making progress!!! :D

Could you clean up some of the files that you're not using anymore. You can add a new commit that deletes them, that way you can go back and get them from the old commits if you change your mind.

**Comment by powerjg** [comment]
Watch what user you're using to create commits. The last couple are owned by "root". Did you edit/commit inside the docker container? You probably shouldn't do that.

File: configs/example/dinocpu.py (hunk 1 above)
Comment (nganjehloo) [reply configs/example/dinocpu.py]: I think I init the memobject in dinocpu.py and pass it as a parameter to DinoCPU in the other config file. Does that not work? I'm still a bit iffy on how the config file passes things around.

**Review by nganjehloo — DISMISSED** [review]

**Comment by powerjg** [comment]
Can you add a commit that deletes all of the files you're not using right now? It's hard to review this PR with a bunch of files that I don't know whether they are being used or not.

Also, you should try not to commit things as root (i.e., in your docker container). You should do all interactions with your file system from outside the container and only use the container to build the .v files.

File: src/SConscript [hunk 8]
```diff
@@ -1378,6 +1378,7 @@ if 'opt' in needed_envs:
             LINKFLAGS = Split(ldflags['opt']),
             disable_partial=disable_partial)
 
+
```
Comment (powerjg) [inline src/SConscript]: Revert this change.

File: src/verilator/VerilatorDinoCPU.cc [hunk 9]
```diff
@@ -0,0 +1,114 @@
+/*# Copyright (c) 2019 The Regents of the University of California
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.cc]: this should be named verilator_dino_cpu.cc

File: src/verilator/VerilatorDinoCPU.hh [hunk 10]
```diff
@@ -0,0 +1,59 @@
+/*# Copyright (c) 2019 The Regents of the University of California
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.hh]: Ditto on naming. Underscores, not camel case

File: src/verilator/VerilatorDinoCPU.hh [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "VTop.h"
+#include "VerilatorMemBlackBox.hh"
+#include "params/VerilatorDinoCPU.hh"
+#include "sim/clocked_object.hh"
+
+class VerilatorDinoCPU : public ClockedObject
+{
+    private:
+        void updateCycle();
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.hh]: Comments would be nice.

File: src/verilator/VerilatorDinoCPU.hh [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+
+#ifndef __VERILATOR_VERILATOR_OBJECT_HH__
+#define __VERILATOR_VERILATOR_OBJECT_HH__
+
+#define VM_TRACE 0
+#define VL_THREADED 0
+
+#include "VTop.h"
+#include "VerilatorMemBlackBox.hh"
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.hh]: Should have full path for gem5 headers (e.g., verilator/VerilatorMemBlackBox.hh) Except the file should be named with underscores ;).

File: src/verilator/VerilatorDinoCPU.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+from m5.params import *
+from VerilatorMemBlackBox import VerilatorMemBlackBox
+from ClockedObject import ClockedObject
+
+class VerilatorDinoCPU(ClockedObject):
+    type = 'VerilatorDinoCPU'
+    cxx_header = "verilator/VerilatorDinoCPU.hh"
+
+    verilatorMem = Param.VerilatorMemBlackBox(NULL, "Reference to "
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: Underscores, not camel case.

File: src/verilator/VerilatorMemBlackBox.hh [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+#define VM_TRACE 0
+#define VL_THREADED 0
+
+#include "VTop_DualPortedMemoryBlackBox.h"
+#include "mem/mem_object.hh"
+#include "params/VerilatorMemBlackBox.hh"
+
+class VerilatorMemBlackBox: public MemObject
+{
+    public:
```
Comment (powerjg) [inline src/verilator/VerilatorMemBlackBox.hh]: Two spaces instead of 4

File: src/verilator/VerilatorMemBlackBox.hh [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        VTop_DualPortedMemoryBlackBox* blkbox;
```
Comment (powerjg) [inline src/verilator/VerilatorMemBlackBox.hh]: 4 spaces instead of 8.

File: src/verilator/VerilatorMemBlackBox.hh [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+        void doFetch();
+        void doMem();
+
+        class VerilatorMemBlackBoxPort : public MasterPort
+        {
+            private:
```
Comment (powerjg) [inline src/verilator/VerilatorMemBlackBox.hh]: Ditto above

File: src/verilator/verilator.py (hunk 7 above)
Comment (powerjg) [reply src/verilator/verilator.py]: Might want to remove this file.

**Review by powerjg — COMMENTED** [review]
Rather than checking in *all* of the verilator generated files, it would be better to just check in the .v file and a README and a Makefile on how to generate the rest. This is urgent, but something to think about at some point. Also, you'll want to add a .gitignore in the ext/verilator directory to ignore all of these generated files.

File: src/verilator/VerilatorDinoCPU.hh [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+        VerilatorMemBlackBox * verilatorMem;
+        EventFunctionWrapper event;
+        Tick clkperiod;
+        int designStages;
+        int cyclesPassed;
+        VTop top;
+    public:
+        VerilatorDinoCPU(VerilatorDinoCPUParams *p);
+        ~VerilatorDinoCPU();
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.hh]: BTW, it's this line that should be removed to get it to compile.

File: src/verilator/README.md [hunk 18]
```diff
@@ -0,0 +1,26 @@
+Basic Verilator Wrapper
```
Comment (powerjg) [inline src/verilator/README.md]: Is this file up to date?

File: src/verilator/VerilatorDinoCPU.py [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+from m5.params import *
+from VerilatorMemBlackBox import VerilatorMemBlackBox
+from ClockedObject import ClockedObject
+
+class VerilatorDinoCPU(ClockedObject):
+    type = 'VerilatorDinoCPU'
+    cxx_header = "verilator/verilator_dino_cpu.hh"
+
+    verilatorMem = Param.VerilatorMemBlackBox("Reference to "
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: Underscores (`verilator_mem`) is the correct style.

File: src/verilator/VerilatorDinoCPU.py [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+        "verilator memory interface")
+    clkperiod = Param.Latency("Clock period of device under test")
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: Shouldn't be needed since this is a clocked object.

File: src/verilator/VerilatorDinoCPU.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    stages = Param.Int(1, "Number of stages in the device under test")
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: Is this used?

File: src/verilator/verilator_dino_cpu.cc [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# Authors: Nima Ganjehloo
+*/
+
+#include "VTop_DualPortedMemory.h"
+#include "VTop_DualPortedMemoryBlackBox.h"
+#include "VTop_Top.h"
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: Are all of these headers required?

This is where we need to improve the interface :). We would rather have the slimmest interface possible.

File: src/verilator/verilator_dino_cpu.cc [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+    void* ptr = aligned_alloc(128, sizeof(VerilatorDinoCPU));
+    return new(ptr) VerilatorDinoCPU(this);
+}
+void
+VerilatorDinoCPU::updateCycle()
+{
+
+    DPRINTF(Verilator, "\nTICKING\n");
+    //has verilator finished running?
+    if (Verilated::gotFinish()|| cyclesPassed >= 10){
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: Shouldn't need the `cyclesPassed` thing. You can call `simulate(10000)` in the python config file to only simulate 10,000 ticks (10 ns).

File: src/verilator/verilator_dino_cpu.cc [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+        inform("Simulation has Completed\n");
+        exitSimLoop("Done Simulating", 1) ;
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: We may not want to exit the sim loop here, but just stop scheduling the next cycle.

File: src/verilator/verilator_dino_cpu.cc [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    DPRINTF(Verilator, "INSTRUCTION FETCH?\n");
+
+    verilatorMem->doFetch();
+    verilatorMem->doMem();
+    //run the device under test here through verilator
+    top.clock = 0;
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: I think we want to factor this out into a general verilator interface file. This isn't specific to DinoCPU, all verilated objects would want to do this.

File: src/verilator/verilator_dino_cpu.cc [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+    top.eval();
+    top.clock = 1;
+    top.eval();
+
+    cyclesPassed += 1;
+
+    schedule(event, nextCycle());
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: `if (!Verilator::getFinished()) schedule(...)`

File: src/verilator/verilator_dino_cpu.cc [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+
+void
+VerilatorDinoCPU::reset(int resetCycles)
```
Comment (powerjg) [inline src/verilator/verilator_dino_cpu.cc]: Factor out?

**Review by powerjg — COMMENTED** [review]
Some comments below. Nothing big :). This is super awesome that you got things working!!!

File: src/verilator/README.md (hunk 18 above)
Comment (nganjehloo) [reply src/verilator/README.md]: nope will update asap

File: src/SConscript (hunk 8 above)
Comment (nganjehloo) [reply src/SConscript]: Sorry I'm a bit confused. Revert what?

File: src/verilator/VerilatorDinoCPU.hh (hunk 12 above)
Comment (nganjehloo) [reply src/verilator/VerilatorDinoCPU.hh]: For some reason this caused the style checker to freak out. It would keep trying to sort my includes, and would tell me it successfully did, but then when I tried to commit again it would complain despite already being sorted correctly according to the checker :/

File: src/verilator/VerilatorDinoCPU.py (hunk 21 above)
Comment (nganjehloo) [reply src/verilator/VerilatorDinoCPU.py]: It is. I was hoping to test out a 5 or 3 stage cpu at some point. Right now its only used for resetting the cpu.

File: src/verilator/verilator_dino_cpu.cc (hunk 22 above)
Comment (nganjehloo) [reply src/verilator/verilator_dino_cpu.cc]: Yeah unfortunately I was getting a lot of errors. I might be able to import VTop__Allcls.h which would automatically have what I need based on the verilog design. Ill try that out.

File: src/verilator/verilator_dino_cpu.cc (hunk 24 above)
Comment (nganjehloo) [reply src/verilator/verilator_dino_cpu.cc]: Will gem5 automatically stop if there are no events scheduled?

File: src/verilator/verilator_dino_cpu.cc (hunk 25 above)
Comment (nganjehloo) [reply src/verilator/verilator_dino_cpu.cc]: Yup. I was thinking that this seperate routine should schedule an updateCycle event after it does what it needs too. I'll change it like that for the time being.

File: src/verilator/verilator_dino_cpu.cc (hunk 27 above)
Comment (nganjehloo) [reply src/verilator/verilator_dino_cpu.cc]: Factor out as in shouldn't be part of the DinoCPU? LIke put into a testbench class of some sort?

File: src/SConscript (hunk 8 above)
Comment (powerjg) [reply src/SConscript]: There's an extra newline, according to the diff here.

File: src/verilator/verilator_dino_cpu.cc (hunk 24 above)
Comment (powerjg) [reply src/verilator/verilator_dino_cpu.cc]: Yes.

File: src/verilator/verilator_dino_cpu.cc (hunk 27 above)
Comment (powerjg) [reply src/verilator/verilator_dino_cpu.cc]: Resetting is something you might want to do for all verilated models, not just the DINO CPU. So, this looks like something that should be in the general interface, not in the DINO CPU specifically.

Comment (nganjehloo) [reply src/verilator/verilator_dino_cpu.cc]: Ahh gotcha.

File: ext/verilator/DualPortedMemoryBlackBox.v [hunk 28]
```diff
@@ -0,0 +1,33 @@
+module DualPortedMemoryBlackBox(
```
Comment (powerjg) [inline ext/verilator/DualPortedMemoryBlackBox.v]: This file needs a copyright statement.

File: ext/verilator/SConscript [hunk 29]
```diff
@@ -0,0 +1,14 @@
+Import('main')
```
Comment (powerjg) [inline ext/verilator/SConscript]: Copyright.

**Review by powerjg — COMMENTED** [review]
You probably don't want to have the libVTop__ALL.a file checked in.

File: src/SConscript [hunk 30]
```diff
@@ -242,10 +242,12 @@ def blobToCpp(data, symbol, cpp_code, hpp_code=None, namespace=None):
     if hpp_code is not None:
         cpp_code('''\
 #include "blobs/{}.hh"
+
```
Comment (powerjg) [inline src/SConscript]: Still some changes in this file that should be dropped. 

BTW, feel free to rebase this branch. No one else is using is IIRC.

File: src/verilator/VerilatorDinoCPU.py [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+
+
+from m5.params import *
+from VerilatorMemBlackBox import VerilatorMemBlackBox
+from ClockedObject import ClockedObject
+
+class VerilatorDinoCPU(ClockedObject):
+  type = 'VerilatorDinoCPU'
+  cxx_header = "verilator/verilator_dino_cpu.hh"
+  stages = Param.Int(1, "Number of stages in the device under test")
```
Comment (powerjg) [inline src/verilator/VerilatorDinoCPU.py]: What is this meant to be used for. I can tell you have some kind of vision here, but I'm not sure what it is :). Maybe we can talk about it in the Thursday meeting or next time I see you in the lab.

**Review by powerjg — COMMENTED** [review]
A couple more comments. I didn't mean to submit the other one...

File: src/verilator/verilator_dino_cpu.cc (hunk 27 above)
Comment (animalvgamer) [reply src/verilator/verilator_dino_cpu.cc]: unsigned int?

File: src/verilator/dpi_manager.cc [hunk 32]
```diff
@@ -45,11 +45,28 @@ VerilatorDinoCPU * dinoCPU = nullptr;
...
+  if (imem_address == 0x80000000 ){
+    DPRINTF(Verilator, "DPI INST FETCH ILLEGAL INST EXCEPTION\n");
+    return 0x00c0006f;//jal x0, 12; effectively halt the processor
+  }else if ( imem_address == 0x80000004 ){
+    DPRINTF(Verilator, "DPI INST FETCH TRAPPING FOR ECALL\n");
+    return 0x0080006f;//jal x0, 8; effectively halt the processor
+  }else if ( imem_address == 0x80000008 ){
+    DPRINTF(Verilator, "DPI INST FETCH TRAPPING FOR EBREAK\n");
+    return 0x0040006f;//jal x0, 4; effectively halt the processor
+  }else if ( imem_address > 0x80000008 ){
```
Comment (ayaz91) [inline src/verilator/dpi_manager.cc]: Can your instructions never be mapped beyond this address?

Comment (nganjehloo) [reply src/verilator/dpi_manager.cc]: they easily can but the processor in question is 32 bit so I was intending for the remaining bytes to be reserved for trap/exception handlers. I currently don't redirect the cpu to such handlers because I don't have any that I've written or available to me so I just stall the processor if a trap or exception is raised. Any ideas on how to handle this better?

**Comment by powerjg** [comment]
When's the next time you're going to be in the office? Let's go through the recent updates in person :).

**Comment by nganjehloo** [comment]
I should be there in 1.5 hrs.

**Comment by powerjg** [comment]
Lol. I didn't expect it would be today. I'm not planning on coming in. Will
you be around tomorrow?

**Comment by nganjehloo** [comment]
Oh right. It's a holiday lol. Yeah i will be in tomorrow as well.

**Comment by powerjg** [comment]
You may want to check out https://github.com/powerjg/gem5/blob/devel/accel which has code for using devices/drivers in SE mode.

File: src/verilator/nvdla/NVDLA.py [hunk 33]
```diff
@@ -33,7 +33,8 @@
 class NVDLAWrapper(DrivenObject):
   type = 'NVDLAWrapper'
   cxx_header = "verilator/nvdla/nvdla_wrapper.hh"
-  do_trace = Param.Bool(True, "Use the specified trace file?")
   buf_clear_cycles = Param.Int(4096,
     "Number of cycles to clear hardware buffers")
-  trace_file = Param.String("input.txn", "Name of trace file to run test" )
+  nvdla_sys = Param.System(Parent.any, "Pointer to nvdla system")
+  axi_2_gem5 = Param.MemObject(Parent.any,
```
Comment (powerjg) [inline src/verilator/nvdla/NVDLA.py]: Will this really be a parent? I suggest not using a proxy here.

File: src/verilator/nvdla/axi_to_mem.hh [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+    public VerilatorMemBlackBox::VerilatorMemBlackBoxPort
+    {
+      public:
+
+         AXIToMemPort(const std::string& name,
+                    AXIToMem *owner) :
+                    AXIToMemPort(name, owner)
+                { }
+
+        vvoid sendTimingPacket(PacketPtr pkt);
```
Comment (powerjg) [inline src/verilator/nvdla/axi_to_mem.hh]: typo.

File: src/verilator/nvdla/NVDLA.py (hunk 33 above)
Comment (nganjehloo) [reply src/verilator/nvdla/NVDLA.py]: Yeah I wasn't really sure what would go here. based off the Proxy Factory code I figured any would go down the tree and eventually hit what I was asking for. Would the system i passed have a way of getting the memobject? I couldn't find any getters in the class declaration

Comment (powerjg) [reply src/verilator/nvdla/NVDLA.py]: No.. you would probably want to do the following in the config file:

```
system.nvdla = NVDLAWrapper()
system.nvdla.axi_2_gem5 = AXIWrapper()
```
I'm not exactly sure what the `AXIWrapper` is here or if you have that done already. But basically, you don't think you want the `nvdla` to be a child of it's memory "port". Instead you want the memory port to be a child of the `nvdla`.

Comment (nganjehloo) [reply src/verilator/nvdla/NVDLA.py]: I see. I think I fixed it.
